include/boost/process/detail/posix/executor.hpp

```diff
@@ -445,8 +453,10 @@ child executor<Sequence>::invoke(boost::mpl::false_, boost::mpl::false_)
     if (_ec)
     {
         //if an error occurred we need to reap the child process
-        ::waitpid(this->pid, nullptr, WNOHANG);
+        int block_mode = _msg == "execve failed" ? 0 : WNOHANG;
```

Comment:
Why is this conditional?

Reply:
Great question. 
**First we must know**:
Doesn't Block - `::waitpid(this->pid, nullptr, WNOHANG); `
Blocks until it has a child zombie to reap - `::waitpid(this->pid, nullptr, 0); `

We don't always want to block. For example, if we look in `_read_error()`, there are several things that can assign a value to `_ec`. One of them is a failed `::read()`. Failed `::read()` doesn't ensure us that we have a zombie child to reap. We want to use  `::waitpid(this->pid, nullptr, 0); ` only when we are sure we have a zombie to reap so we won't be blocked for a long time.
Only if `_msg == "execve failed"` we know the child has exited or will exit in its next few lines.

Reply:
I don't think that's correct. The closed pipe would indicate that the child process exited already, so WNOHANG should be correct.

Reply:
And comparing error messages is just bad practice. Why can't we just wait and ignore the error if the child doesn't exists?